### PR TITLE
Add Rails/OutputSafety Cop

### PR DESCRIPTION
**Pull request by josh**
This adds a new cop tighten security around Rails output safety helpers like `.html_safe` and `raw`. While you may need `.html_safe` in a handful of cases, it was too easy to casually use without careful review. Instead of just using `content_tag(:div, something)` in a safe way in a helper method, `"<div>#{something}</div>".html_safe` would be used instead. In instances `.html_safe` were is necessary, we use an inline `# rubocop:disable Rails/OutputSafety` comment after security review.

### CODE REVIEW

File: lib/rubocop/cop/rails/output_safety.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# encoding: utf-8
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # This cop checks for the use of output safety calls like html_safe and
+      # raw.
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/output_safety.rb]: Add here some examples of code the cop would consider offensive. See other cops for reference.

File: lib/rubocop/cop/rails/output_safety.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      class OutputSafety < Cop
+        MSG = 'Prefer `safe_join` or other tag helpers instead'.freeze
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/output_safety.rb]: I think this message can made a bit more informative.

File: config/disabled.yml [hunk 3]
```diff
@@ -5,6 +5,10 @@
 Rails:
   Enabled: false
 
+Rails/OutputSafety:
+  Description: 'The use of `html_safe` or `raw` maybe a security risk.'
```
Comment (bbatsov) [inline config/disabled.yml]: maybe -> may be

File: config/disabled.yml [hunk 4]
```diff
@@ -5,6 +5,10 @@
...
+  Enabled: false
```
Comment (bbatsov) [inline config/disabled.yml]: Why is this disabled by default?

**Comment by bbatsov** [comment]
You might want to add some rule about this in the Rails style guide document as well. Such changes require a changelog entry.

File: config/disabled.yml (hunk 4 above)
Comment (josh) [inline config/disabled.yml]: It seemed pretty opinionated by default. There are going to be some places you'll need to use this. For our app, I'm just taking the stance that all those places should use an explicit `# rubocop:disable` comment to whitelist those few spots.

**Comment by josh** [comment]
> You might want to add some rule about this in the Rails style guide document as well

Wheres that at?

> Such changes require a changelog entry.

👍

**Comment by josh** [comment]
@bbatsov pushed some changes based on your suggestions.

**Comment by bbatsov** [comment]
> Wheres that at?

https://github.com/bbatsov/rails-style-guide

Apart from the failing build, the cops looks OK to me. You'll have to rebase, though. What's the reason why this is disabled by default? I'm guessing it produces some false positives, right?

**Comment by josh** [comment]
> Apart from the failing build, the cops looks OK to me. You'll have to rebase, though.

👍 I'm on it.

> What's the reason why this is disabled by default? I'm guessing it produces some false positives, right?

I'll just move it to enabled by default.

**Comment by josh** [comment]
@bbatsov updated with changes.

**Comment by bbatsov** [comment]
👍 Thanks!

**Comment by lucascaton** [comment]
Hi @josh,

Any suggestions for the following case?

```ruby
= f.submit '&#9654;'.html_safe
```

Cheers.

**Comment by mikegee** [comment]
How about using the decoded character?

```
= f.submit '▶'
```

**Comment by lucascaton** [comment]
Thanks @mikegee, but I was wondering if using a non-ASCII character wouldn't be a bad practice.

**Comment by josh** [comment]
If you really **need** encoded entities, consider something like `= f.submit html_encode('▶')`. Others reading your views have no idea what `&#9654;` means.

**Comment by lucascaton** [comment]
Thanks @josh!

Did you mean `html_escape`? Because either `html_escape` or `h` compiles to the following HTML:
```html
<input type="submit" name="commit" value="▶" />
```

While `= f.submit '&#9654;'.html_safe` generates a non-ASCII character (which is considered a good practice afaik):

```html
<input type="submit" name="commit" value="&#9654;" />
```

Cheers.

**Comment by lastobelus** [comment]
I think the cop should allow raw() for uninterpolated strings.

**Comment by mdavidn** [comment]
Butting heads against this cop with `'&nbsp;'.html_safe`. There's no Rails helper for HTML entities. Uninterpolated strings should not be a violation, as there's no risk of code injection.

```rb
safe_join(words, '&nbsp;'.html_safe)
```

**Pull request merged**